#SuperPhy

![LICENSE](https://i.creativecommons.org/l/by-sa/4.0/88x31.png)
#This work is licensed under a [Creative Commons Attribution-ShareAlike 4.0 International License](http://creativecommons.org/licenses/by-sa/4.0/).

#Description / Abstract
Predictive genomics involves the translation of raw genome sequence data into an assessment of the phenotypes exhibited by the organism from which the genome came. These phenotypes can range from how well the organism is able to survive in a particular environment to whether the organism will cause severe human disease. Advances in DNA sequencing technology have created new opportunities in predictive genomics, which has applications in fields as diverse as clinical medicine and epidemiology, where performing real-time, genome-based surveillance and identification of phenotypic characteristics of pathogens may be life-saving and reduce costs to the health care system through improved diagnostic, prevention and treatment efforts.

New analytical tools and infrastructure are needed to analyze the raw genome data generated by current sequencing technology, a task that is recognized as a challenge worldwide. Significant progress has been made in the development of generic tools that are broadly applicable to all microorganisms; however, a fundamental missing component is the ability to analyze these data in the context of organism-specific knowledge. This accumulated knowledge from decades of research can provide a meaningful interpretation and context for the data. 

We have implemented an online predictive genomics platform [SuperPhy](http://lfz.corefacility.ca/superphy/) for _Escherichia coli_. The platform integrates the analyses tools and genome sequence data for all publicly available genomes with metadata and facilitates the upload of genome sequences from users under public or private settings. The predictive genomics platform provides near real-time analyses of thousands of genome sequences using novel computational approaches with results that are understandable and useful to a wide community, including those in the fields of clinical medicine, epidemiology, ecology, and evolution. Specific analyses include identification of: 
    - virulence and antimicrobial resistance determinants 
    - epidemiological associations between specific genotypes, biomarkers, geospatial distribution, host, source, and other meta-data 
    - statistically significant clade-specific genome markers (presence / absence of specific genomic regions, and single-nucleotide polymorphisms) in bacterial populations
    - _in silico_ molecular type for traditional wet-lab typing methods including Shiga-toxin subtyping
