Selection Criteria:

Research that has already delivered on intermediate or long-term outcomes of the GRDI, with impact (outcomes) that can be described
Focus on end-user applications of knowledge developed through GRDI investment
Identified end-user/stakeholder who can be interviewed


Information on GRDI achievement:
Title: SuperPhy: Predictive genomics for priority food-borne pathogens

Executive summary (short outline):
Predictive genomics involves the translation of raw genome sequence data into an assessment of the phenotypes exhibited by the organism from which the genome came. These phenotypes can range from how well the organism is able to survive in a particular environment, such as high salt concentration, to whether the organism will cause severe human disease. Advances in DNA sequencing technology have created new opportunities in the field of predictive genomics, which has application to fields as diverse as clinical medicine and epidemiology, where performing real-time, genome-based surveillance and identification of phenotypic characteristics of pathogens may be life-saving and reduce costs to the health care system through improved diagnostic, prevention and treatment efforts.

New analytical tools and infrastructure are needed to analyze the raw genomic data generated by current sequencing technology, a task that is recognized as a challenge worldwide. Significant progress has been made in the development of generic tools that are broadly applicable to all microorganisms; however, a fundamental missing component is the ability to analyze this data in the context of organism-specific knowledge. This accumulated knowledge from decades of research can provide a meaningful interpretation and context for the data. 

We have implemented an online predictive genomics platform (http://lfz.corefacility.ca/superphy/) for Escherichia coli. The platform integrates the analyses tools and genome sequence data for all publicly available genomes and facilitates the upload of genome sequences from users under public or private settings. The predictive genomics platform provides near real-time analyses of thousands of genome sequences using novel computational approaches with results that are understandable and useful to a wide community, including those in the fields of clinical medicine, epidemiology, ecology, and evolution. Specific analyses include identification of: 1) virulence and antimicrobial resistance determinants 2) epidemiological associations between specific genotypes, biomarkers, geospatial distribution, host, source, and other meta-data; 3) statistically significant clade-specific genome markers (presence / absence of specific genomic regions, and single-nucleotide polymorphisms) in bacterial populations; 4) in silico molecular type for traditional wet-lab typing methods including Shiga-toxin subtyping.


Brief description of research, years funded under the GRDI
The project is funded under the GRDI round 6 competition as `'BioTools for the predictive genomics of priority foodborne pathogens'. This new project originated from work done in other GRDI-funded projects, names the Food and Water Safety program and our own intramural GRDI 5 project.

What issue/opportunity did this research seek to address?
The ability to analyze whole-genome sequence data in the context of organism-specific knowledge is not currently possible.  There are many developed tools that are broadly applicable to all microorganisms, but none that provide expert-guided species-specific knowledge.

How did this research address the problem?
By implementing an online predictive genomics platform (http://lfz.corefacility.ca/superphy/) for Escherichia coli. The platform integrates the analyses tools and genome sequence data, as well as the accumulated knowledge from decades of research that can be used to provide a meaningful interpretation and context for the data. 

What are the impacts/outcomes achieved from this research?
The ability for near real-time analyses of thousands of genome sequences using novel computational approaches with results that are understandable and useful to a wide community, including those in the fields of clinical medicine, epidemiology, ecology, and evolution.

Who benefits from the outcomes of the research?
Basic researchers,  those in the fields of clinical medicine, epidemiology, ecology, evolution, and eventually the aim is to have healthier citizens as the accumulated knowledge is used to mitigate illness associated with sporadic and outbreak cases of foodborne illness caused by bacterial pathogens.

Are any particular regions that benefit from the results more than others?
Not that we can tell.

Where to go from here; does this research continue to be funded under the GRDI?
Yes, the proof of concept using Escherichia coli has been established, but under the GRDI6 grant, additional knowledge regarding E. coli will be added, and predictive genomics for the two additional species Salmonella enterica and Campylobacter jejunii will be added.

Illustrative photo/diagram




Name of organizations involved (lead; collaborators)    
Public Health Agency of Canada


Name and coordinates of primary contact
Dr. Vic Gannon



Name and coordinates of stakeholder/end-user who can be interviewed

For reference: intermediate and long-term outcomes described in the GRDI Performance Measurement Framework for Phase V:

Intermediate outcomes:
Government policy makers and regulators have used research results for evidence-based regulatory, policy, and resource management decisions
Private and public stakeholders involved in the innovation continuum in Canada have adopted innovative or improved tools and processes using research results
Long-term outcomes:
Improved human health in Canada
Enhanced sustainability and management of Canadaâ€™s environment, agriculture, forestry and fisheries sectors
Improved food safety and security in Canada
